share/wake/lib/core/list.wake

```diff
@@ -802,3 +805,76 @@ export def groupBy (cmp: a => a => Boolean) (list: List a): List (List a) =
             (elem, head, rest), tail
         _ = (elem, Nil), acc
     foldr combo Nil sorted
+
+export def fromOption (x: Option a): List a =
+    getOrElse Nil $ omap (_, Nil) x
+
+# TODO: replace at with atT
+export def atT (i: Integer) (list: List a): List a =
+    at i list | fromOption
+
+export def atE (i: Integer): (a => a) => List a => List a =
+    def edit f l =
+      require Pair first (x, follow) = splitAt i l else l
+      first ++ (f x, follow)
+    edit
+
+export def headT: (List a) => (List a) = atT 0
+
+export def headE: (a => a) => List a => List a = atE 0
+
+export def takeE (i: Integer): (a => a) => List a => List a =
+  def edit f l =
+    def Pair first follow = splitAt i l
+    (map f first) ++ follow
+  edit
+
+# NOTE: This is an easier-to-use version of `find`
+# Do we care to keep it?
+export def findFirst (p: a => Boolean): List a => List a =
+    find p _ | fromOption | map getPairFirst
+
+export def findFirstE (p: a => Boolean): (a => a) => List a => List a =
+    def edit f l =
+      require (x, xs) = l
+      if p x then
+        f x, xs
+      else
+        x, edit f xs
+
+    edit
+
+export def filterE (p: a => Boolean): (a => a) => List a => List a =
```

Comment:
Rename this to `editFilter` and restructure it a bit:
```
export def editFilter (selectFn: a => Boolean) (editFn: a => a): List a => List a =
  def editIf x = if selectFn x then editFn x else x
  map editIf
```
(you want to be sure that the `editFn` inlines into `editIf` and in turn, `map`'s loop)